pkg/agent/nodeportlocal/npl_agent_init.go

```diff
@@ -33,10 +34,184 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
 )
 
 const resyncPeriod = 60 * time.Minute
 
+func updatePodNodePortCache(cache map[string][]k8s.PodNodePort, podIP string, nodePort, podPort int) {
+	v := k8s.PodNodePort{
+		NodePort: nodePort,
+		PodPort:  podPort,
+	}
+	_, exists := cache[podIP]
+	if !exists {
+		cache[podIP] = []k8s.PodNodePort{v}
+		return
+	}
+	cache[podIP] = append(cache[podIP], v)
+}
+
+// rulesToPodPortMap is responsible for going through the NPL rules and figures out if some of these
+// rules are still needed. To check if a rule is needed, it first checks if the relevant Pod exists,
+// if not, it deletes the rule. This function adds the port rules to the portTable and responds with
+// a list of visited Pods.
+func rulesToPodPortMap(portTable *portcache.PortTable, kubeClient clientset.Interface) (map[string][]k8s.PodNodePort, error) {
+	podPortRules, err := portTable.PodPortRules.GetAllRules()
+	if err != nil {
+		klog.Errorf("error in fetching the Pod port rules: %s", err.Error())
+		return nil, errors.New("error in fetching the Pod port rules: " + err.Error())
+	}
+
+	podNodePortCache := make(map[string][]k8s.PodNodePort)
+
+	for k, v := range podPortRules {
+		// k: NodePort
+		// v: {PodIP,Pod port}
+		if k > portTable.EndPort || k < portTable.StartPort {
+			// delete this rule, as the Node port falls in an invalid range
+			err := portTable.PodPortRules.DeleteRule(k, v.PodIP)
+			if err != nil {
+				klog.Errorf("error in deleting the rule for port %d and Pod IP %s", k, v.PodIP)
+				continue
+			}
+		}
+		podList, err := kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
```

Comment:
This won't scale well and may take some time to finish the initialization because:
1. kube-apiserver doesn't have index for podIP, it will linear search to get the target Pod.
2. each pod will do the same on apiserver.

I think you can list local Pods once and build a podIP to pod map then use it in this loop.
Or you can leverage the PodLister added in https://github.com/vmware-tanzu/antrea/pull/1743.

Reply:
Noted. Yeah, will change it.